x/wasm/internal/keeper/keeper_test.go

```diff
@@ -431,6 +433,43 @@ func TestInstantiateWithNonExistingCodeID(t *testing.T) {
 	require.Nil(t, addr)
 }
 
+func TestInstantiateWithCallbackToContract(t *testing.T) {
+	ctx, keepers := CreateTestInput(t, false, SupportedFeatures, nil, nil)
+	var (
+		wasmerMock = &selfCallingInstMockWasmer{}
+		err        error
+	)
+	keepers.WasmKeeperRef.wasmer = wasmerMock
+	keepers.WasmKeeper.wasmer = wasmerMock
+	example := StoreHackatomExampleContract(t, ctx, keepers)
+	_, err = keepers.WasmKeeper.Instantiate(ctx, example.CodeID, example.CreatorAddr, nil, nil, "test", nil)
+	require.NoError(t, err)
+	assert.True(t, wasmerMock.excuteCalled)
+}
+
+// mock to call itself on instantiation
+type selfCallingInstMockWasmer struct {
+	AlwaysFailMockWasmer
```

Comment:
That always failed when you override Instantiate and Execute is a bit confusing. I expect this to be an error, not actually do something. Mayb just provide a Query implementation (or extend NoopMockWasmer with a positive query operation).

Reply:
The `AlwaysFailMockWasmer` implements all the methods to satisfy the WasmEngine interface. I overwrite only 3 methods here for the test so that I can create the proper data. A call to any other method should fail the test.
An always panic wasmer engine implementation would be more accurate though....
I will refactor this.